url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/expert_system/Tonality.java

```diff
@@ -0,0 +1,173 @@
+// ##################################################################
+// ##################################################################
+// # Tonality Class: Used to store data about the current			#
+// #	 	         tonality (Key and major/minor inclination). 	#
+// #															    #	
+// #	Stores the current tonality (eg C major), Its related keys	#
+// #	and their related chord type. Each Chord type allows access	#
+// #			 to the scales playable over that chord.			#
+// ##################################################################
+// ##################################################################
+package expert_system;
+
+public class Tonality {
+	
+	// ##########################################################
+	// #					      FIELDS						#
+	// ##########################################################
+
+	public Key root;
+	public Key [] relatedKeys;
+	
+	public int Major_Minor_Tonality;
+	
+	// ##########################################################
+	// #					   CONSTRUCTORS						#
+	// ##########################################################
+
+	public Tonality(Key k, int i) {
+		
+		root = k;
+		Major_Minor_Tonality = i;
+		
+		if(Major_Minor_Tonality == 1)
+			root.setChord(Chord.MAJOR);
+		else
+			root.setChord(Chord.MINOR_DORIAN);
+		
+		this.setRelatedKeys();
+	}
+
+	// ##########################################################
+	// #	METHODS: Getters/Setters for relatedKeys, 			#
+	// #			 Major_Minor_Tonality, root, 				#
+	// #			 relativeMajor_Minor						#
+	// ##########################################################
+	
+	// Returns relatedKeys field
+	public Key[] getRelatedKeys() {
+		return relatedKeys;
+	}
+	// Sets relatedKeys field
+	public void setRelatedKeys() {
+		
+		// Switch statement to determine whether its major or minor
+		switch(Major_Minor_Tonality){
+		
+		// If Minor set the relatedKeys to the intervals in the Aeolian mode
+		// (Pure minor), and set each of their Chord's to the chord related
+		// to that interval - eg 5th -> Dominant 7th.
+		case 0:
+			Key [] temp = root.getAllKeys();
+			
+			try{
+				relatedKeys = new Key[root.minorIntervals.length];
+				for(int k = 0; k < root.getAllKeys().length; k++){
+					if(temp[k] == root){
+						int itr = 0;
+						for(int i = 0; i < root.minorIntervals.length; i++){
+							
+							itr += root.minorIntervals[i];
+							this.relatedKeys[i] = temp[k+itr];
+							
+						}
+						break;	
+					}	
+				}
+			}catch(Exception e){
+				
+				e.printStackTrace();
+				
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 63:81, 97:115.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.